R/removeEnvironments.R

```diff
@@ -0,0 +1,169 @@
+#' @title removeEnvironments - remove environments from ggplot2 objects
+#' @param x a plot object. Currently, only ggplot2 objects and jaspGraphsPlot are supported, other objects are returned unchanged
+#' @param debugInfo Should debugInfo be printed? Possible values are 0, no info is shown; 1, some info is shown; or 2, a lot of info is shown.
+#' @param ... internal.
+#'
+#' @details This function may breaks ggplot2 objects that rely on looking up values in environments.
+#' see also https://github.com/tidyverse/ggplot2/issues/3994 and https://github.com/tidyverse/ggplot2/issues/4056
+#'
+#' @example inst/examples/ex-removeEnvironments.R
+#' @returns an object of the same class as the input
+#' @export
+removeEnvironments <- function(x, debugInfo = 0,...) {
+  UseMethod("removeEnvironments", x)
+}
+
+#' @export
+removeEnvironments.ggplot <- function(x, debugInfo = 0, ...) {
+  x$plot_env <- rlang::new_environment()
+  return(removeEnvironmentsInner(x, debugInfo = debugInfo, ...))
+}
+
+#' @export
+removeEnvironments.jaspGraphsPlot <- function(x, debugInfo = 0, ...) {
+
+  for (i in seq_along(x)) {
+    if (debugInfo >= 1) cat("subplot", i, "\n")
+    x[[i]] <- removeEnvironmentsInner(x, debugInfo = debugInfo, ...)
+  }
+  return(x)
+}
+
+#' @export
+removeEnvironments.default <- function(x, debugInfo = 0, ...) {
+  if (debugInfo >= 1) cat("dispatched to removeEnvironments.default\n")
+  return(x)
+}
+
+removeEnvironmentsInner <- function(
+    x,
+    exitEarlyClasses = c("unit", "element", "data.frame", "FacetNull"),
+    replacementEnv = rlang::new_environment(parent = baseenv()),
+    environmentsAlreadyVisited = rlang::new_environment(),
+    debugInfo = 1,
+    level = deparse1(substitute(x))) {
+
+  if (inherits(x, exitEarlyClasses) || isNamespace(x))
+    return(x)
+
+  if (debugInfo >= 2) {
+    cat("level:", level, "\n")
+    cat("class:", paste(class(x), collapse = ", "), "\n")
+  }
+
+  if (inherits(x, "uneval")) { # returned by ggplot2::aes
+
+    if (debugInfo >= 1) {
+      cat("level:", level, "\n")
+      cat("class:", paste(class(x), collapse = ", "), "\n")
+      cat("replacing environment of subelements of x\n")
+    }
+
+    for (i in seq_along(x)) {
+      environment(x[[i]]) <- replacementEnv
```

Comment:
Now that I think of it, this part (and the other parts that remove environments) actually remove *all* environments. Perhaps I should add an option to only remove specified environments (by default `ggplotObj$plot_env`).